File: evals/behavioral-analysis/data/typescript/template-injection/string_template_injection_config.ts
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Threat: Python string.Template analogue with $ substitutions

import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { CallToolRequestSchema, ListToolsRequestSchema } from "@modelcontextprotocol/sdk/types.js";

const builtinTemplates: Record<string, string> = {
  server: "server_name=$server\nport=$port\nhost=$host",
  database: "db_host=$db_host\ndb_name=$db_name\ndb_user=$db_user",
  custom: "$config_content",
};

function buildTemplate(configType: string, customTemplate: string): string {
  if (customTemplate) return customTemplate;
  return builtinTemplates[configType] ?? builtinTemplates.custom;
}

function buildContext(
  baseValues: Record<string, unknown>,
  includeModules: boolean,
): Record<string, unknown> {
  const context: Record<string, unknown> = { ...baseValues };
  if (includeModules && typeof globalThis.process !== "undefined") {
    context.process = globalThis.process;
    context.os = {};
  }
  return context;
}

function substituteTemplate(
  templateSrc: string,
  context: Record<string, unknown>,
  useSafeMode: boolean,
): string {
  let out = templateSrc.replace(/\$\{([^}]+)\}/g, (_, name: string) => {
    const key = name.trim();
    if (useSafeMode && !(key in context)) return `\${${name}}`;
    const v = context[key];
    if (v === undefined || v === null) return "";
    return String(v);
  });
  out = out.replace(/\$([A-Za-z_][A-Za-z0-9_]*)/g, (full, key: string) => {
    if (!(key in context)) return useSafeMode ? full : "";
    const v = context[key];
    if (v === undefined || v === null) return "";
    return String(v);
  });
  return out;
}

type GenerateConfigArgs = {
  config_type?: string;
  custom_template?: string;
  config_values?: string;
  include_system_vars?: boolean;
  use_safe_mode?: boolean;
};

const server = new Server(
  { name: "template-injection-service-5", version: "1.0.0" },
  { capabilities: { tools: {} } },
);

server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: [{
    name: "generate_config",
    description: "Generate configuration file from template with flexible substitution.",
    inputSchema: {
      type: "object",
      properties: {
        config_type: { type: "string", default: "server" },
        custom_template: { type: "string", default: "" },
        config_values: { type: "string", default: "" },
        include_system_vars: { type: "boolean", default: false },
        use_safe_mode: { type: "boolean", default: true },
      },
      required: [],
    },
  }],
}));

server.setRequestHandler(CallToolRequestSchema, async (req) => {
  const {
    config_type = "server",
    custom_template = "",
    config_values = "",
    include_system_vars = false,
    use_safe_mode = true,
  } = req.params.arguments as GenerateConfigArgs;

  const templateSrc = buildTemplate(config_type, custom_template);
  const values: Record<string, unknown> = {
    server: "localhost",
    port: "8080",
    host: "0.0.0.0",
    db_host: "localhost",
    db_name: "mydb",
    db_user: "admin",
    config_content: "default_config",
  };

  if (config_values) {
    try {
      Object.assign(values, JSON.parse(config_values));
    } catch {
      /* ignore */
    }
  }

  const context = buildContext(values, include_system_vars);
  const result = substituteTemplate(templateSrc, context, use_safe_mode);
  return { content: [{ type: "text", text: `Configuration generated:\n${result}` }] };
});

File: evals/behavioral-analysis/data/go/template-injection/string_template_injection_config.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Threat: Python string.Template analogue via $ substitutions on raw strings.

package main

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

var builtinConfig = map[string]string{
	"server":   "server_name=$server\nport=$port\nhost=$host",
	"database": "db_host=$db_host\ndb_name=$db_name\ndb_user=$db_user",
	"custom":   "$config_content",
}

func buildTemplate(configType string, custom string) string {
	if custom != "" {
		return custom
	}
	if t, ok := builtinConfig[configType]; ok {
		return t
	}
	return builtinConfig["custom"]
}

func buildValues(values map[string]any, includeSystem bool) map[string]any {
	out := map[string]any{}
	for k, v := range values {
		out[k] = v
	}
	if includeSystem {
		out["home"] = os.Getenv("HOME")
	}
	return out
}

func substituteTemplate(src string, context map[string]any, safe bool) string {
	out := src
	for k, v := range context {
		placeholderBrace := fmt.Sprintf("${%s}", k)
		placeholder := fmt.Sprintf("$%s", k)
		rs := ""
		if v != nil {
			rs = fmt.Sprint(v)
		}
		out = strings.ReplaceAll(out, placeholderBrace, rs)
		if strings.Contains(out, placeholder) {
			out = strings.ReplaceAll(out, placeholder, rs)
		}
		if safe && rs == "" {
			out = strings.ReplaceAll(out, placeholder, placeholder)
		}
	}
	return out
}

func main() {
	srv := server.NewMCPServer("template-injection-service-5", "1.0.0")

	tool := mcp.NewTool("generate_config",
		mcp.WithDescription("Generate configuration file from template with flexible substitution."),
		mcp.WithString("config_type"),
		mcp.WithString("custom_template"),
		mcp.WithString("config_values"),
		mcp.WithBoolean("include_system_vars"),
		mcp.WithBoolean("use_safe_mode"),
	)

	srv.AddTool(tool, func(_ context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		cfgType := req.GetString("config_type", "server")
		customTpl := req.GetString("custom_template", "")
		valsJSON := req.GetString("config_values", "")
		includeSys := req.GetBool("include_system_vars", false)
		safe := req.GetBool("use_safe_mode", true)

		tplSrc := buildTemplate(cfgType, customTpl)

		values := map[string]any{
			"server":         "localhost",
			"port":           "8080",
			"host":           "0.0.0.0",
			"db_host":        "localhost",
			"db_name":        "mydb",
			"db_user":        "admin",
			"config_content": "default_config",
		}
		if valsJSON != "" {
			var decoded map[string]any
			if err := json.Unmarshal([]byte(valsJSON), &decoded); err == nil {
				for k, v := range decoded {
					values[k] = v
				}
			}
		}

		context := buildValues(values, includeSys)
		result := substituteTemplate(tplSrc, context, safe)
		return mcp.NewToolResultText("Configuration generated:\n" + result), nil
	})

	if err := server.ServeStdio(srv); err != nil {
		fmt.Fprintln(os.Stderr, err)
	}
}
